### ISSUE TITLE
HTTP2 async failure on a stream can lead to the connection being aborted

### ISSUE THREAD

**Issue by lorban** [eclipse/jetty.project#11854]
**Jetty version(s)**
12.0.9

**Description**
If an HTTP2 stream gets interrupted by an async failure (like a RESET frame), there is a possibility that the connection gets aborted abruptly.

The cause of this problem is that if the server is busy writing some byte buffers to the socket while `HttpChannel.onFailure()` is called by a different thread, the callback passed to `ChannelResponse.write()` is failed, releasing the byte buffers back to the pool after resetting them while the `SocketChannel.write()` is still running. Upon returning, the NIO call fails with `IllegalArgumentException` while trying to update the `limit`, like so:

```
java.lang.IllegalArgumentException: newPosition > limit: (14 > 0)
	at java.base/java.nio.Buffer.createPositionException(Buffer.java:352)
	at java.base/java.nio.Buffer.position(Buffer.java:327)
	at java.base/java.nio.ByteBuffer.position(ByteBuffer.java:1551)
	at java.base/java.nio.MappedByteBuffer.position(MappedByteBuffer.java:328)
	at java.base/java.nio.MappedByteBuffer.position(MappedByteBuffer.java:73)
	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:237)
	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:158)
	at java.base/sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:574)
	at java.base/java.nio.channels.SocketChannel.write(SocketChannel.java:660)
	at org.eclipse.jetty.io.SocketChannelEndPoint.flush(SocketChannelEndPoint.java:111)
	at org.eclipse.jetty.io.WriteFlusher.flush(WriteFlusher.java:422)
	at org.eclipse.jetty.io.WriteFlusher.write(WriteFlusher.java:275)
	at org.eclipse.jetty.io.WriteFlusher.write(WriteFlusher.java:254)
	at org.eclipse.jetty.io.AbstractEndPoint.write(AbstractEndPoint.java:368)
	at org.eclipse.jetty.http2.internal.HTTP2Flusher.process(HTTP2Flusher.java:292)
	at org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:250)
	at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:231)
	at org.eclipse.jetty.http2.HTTP2Session.frame(HTTP2Session.java:817)
	at org.eclipse.jetty.http2.HTTP2Session.frames(HTTP2Session.java:790)
	at org.eclipse.jetty.http2.HTTP2Stream.send(HTTP2Stream.java:146)
	at org.eclipse.jetty.http2.server.internal.HttpStreamOverHTTP2.sendHeaders(HttpStreamOverHTTP2.java:395)
	at org.eclipse.jetty.http2.server.internal.HttpStreamOverHTTP2.send(HttpStreamOverHTTP2.java:286)
	at org.eclipse.jetty.server.HttpStream$Wrapper.send(HttpStream.java:179)
	at org.eclipse.jetty.server.internal.HttpChannelState$ChannelResponse.write(HttpChannelState.java:1263)
	at org.eclipse.jetty.server.Response$Wrapper.write(Response.java:751)
	at org.eclipse.jetty.server.handler.EventsHandler$EventsResponse.write(EventsHandler.java:353)
	at org.eclipse.jetty.server.Response$Wrapper.write(Response.java:751)
	at org.eclipse.jetty.server.handler.EventsHandler$EventsResponse.write(EventsHandler.java:353)
	at org.eclipse.jetty.server.Response$Wrapper.write(Response.java:751)
	at org.eclipse.jetty.server.handler.ContextResponse.write(ContextResponse.java:56)
	at org.eclipse.jetty.ee10.servlet.ServletContextResponse.write(ServletContextResponse.java:288)
	at org.eclipse.jetty.ee10.servlet.HttpOutput.channelWrite(HttpOutput.java:205)
	at org.eclipse.jetty.ee10.servlet.HttpOutput.complete(HttpOutput.java:436)
	at org.eclipse.jetty.ee10.servlet.ServletContextResponse.completeOutput(ServletContextResponse.java:212)
	at org.eclipse.jetty.ee10.servlet.ServletChannel.handle(ServletChannel.java:576)
	at org.eclipse.jetty.ee10.servlet.ServletHandler.handle(ServletHandler.java:464)
	at org.eclipse.jetty.server.handler.ContextHandler.handle(ContextHandler.java:851)
	at org.eclipse.jetty.server.Handler$Wrapper.handle(Handler.java:740)
	at org.eclipse.jetty.server.handler.EventsHandler.handle(EventsHandler.java:81)
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:151)
	at org.eclipse.jetty.server.Handler$Wrapper.handle(Handler.java:740)
	at org.eclipse.jetty.server.handler.EventsHandler.handle(EventsHandler.java:81)
	at org.eclipse.jetty.server.Server.handle(Server.java:179)
	at org.eclipse.jetty.server.internal.HttpChannelState$HandlerInvoker.run(HttpChannelState.java:635)
	at org.eclipse.jetty.util.thread.Invocable$ReadyTask.run(Invocable.java:105)
	at org.eclipse.jetty.http2.server.internal.HttpStreamOverHTTP2$1.run(HttpStreamOverHTTP2.java:133)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:478)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:441)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:293)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.produce(AdaptiveExecutionStrategy.java:195)
	at org.eclipse.jetty.http2.HTTP2Connection.produce(HTTP2Connection.java:211)
	at org.eclipse.jetty.http2.server.internal.HTTP2ServerConnection.onOpen(HTTP2ServerConnection.java:115)
	at org.eclipse.jetty.io.AbstractEndPoint.upgrade(AbstractEndPoint.java:435)
	at org.eclipse.jetty.server.internal.HttpConnection$HttpStreamOverHTTP1.upgrade(HttpConnection.java:1495)
	at org.eclipse.jetty.server.internal.HttpConnection.upgrade(HttpConnection.java:582)
	at org.eclipse.jetty.server.internal.HttpConnection$HttpStreamOverHTTP1.headerComplete(HttpConnection.java:1321)
	at org.eclipse.jetty.server.internal.HttpConnection$RequestHandler.headerComplete(HttpConnection.java:962)
	at org.eclipse.jetty.http.HttpParser.parseFields(HttpParser.java:1345)
	at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:1641)
	at org.eclipse.jetty.server.internal.HttpConnection.parseRequestBuffer(HttpConnection.java:568)
	at org.eclipse.jetty.server.internal.HttpConnection.onFillable(HttpConnection.java:391)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:322)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:99)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at org.eclipse.jetty.util.thread.MonitoredQueuedThreadPool$1.run(MonitoredQueuedThreadPool.java:73)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:979)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1209)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1164)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

or

```
java.lang.IllegalArgumentException: newPosition > limit: (370 > 0)
	at java.base/java.nio.Buffer.createPositionException(Buffer.java:352)
	at java.base/java.nio.Buffer.position(Buffer.java:327)
	at java.base/java.nio.ByteBuffer.position(ByteBuffer.java:1551)
	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:237)
	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:158)
	at java.base/sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:574)
	at java.base/java.nio.channels.SocketChannel.write(SocketChannel.java:660)
	at org.eclipse.jetty.io.SocketChannelEndPoint.flush(SocketChannelEndPoint.java:111)
	at org.eclipse.jetty.io.WriteFlusher.flush(WriteFlusher.java:422)
	at org.eclipse.jetty.io.WriteFlusher.write(WriteFlusher.java:275)
	at org.eclipse.jetty.io.WriteFlusher.write(WriteFlusher.java:254)
	at org.eclipse.jetty.io.AbstractEndPoint.write(AbstractEndPoint.java:368)
	at org.eclipse.jetty.http2.internal.HTTP2Flusher.process(HTTP2Flusher.java:292)
	at org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:250)
	at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:231)
	at org.eclipse.jetty.http2.HTTP2Session.frame(HTTP2Session.java:817)
	at org.eclipse.jetty.http2.HTTP2Session.frames(HTTP2Session.java:790)
	at org.eclipse.jetty.http2.HTTP2Stream.send(HTTP2Stream.java:146)
	at org.eclipse.jetty.http2.server.internal.HttpStreamOverHTTP2.sendHeaders(HttpStreamOverHTTP2.java:395)
	at org.eclipse.jetty.http2.server.internal.HttpStreamOverHTTP2.send(HttpStreamOverHTTP2.java:286)
	at org.eclipse.jetty.server.HttpStream$Wrapper.send(HttpStream.java:179)
	at org.eclipse.jetty.server.internal.HttpChannelState$ChannelResponse.write(HttpChannelState.java:1263)
	at org.eclipse.jetty.server.Response$Wrapper.write(Response.java:751)
	at org.eclipse.jetty.server.handler.EventsHandler$EventsResponse.write(EventsHandler.java:353)
	at org.eclipse.jetty.server.Response$Wrapper.write(Response.java:751)
	at org.eclipse.jetty.server.handler.EventsHandler$EventsResponse.write(EventsHandler.java:353)
	at org.eclipse.jetty.server.Response$Wrapper.write(Response.java:751)
	at org.eclipse.jetty.server.handler.ContextResponse.write(ContextResponse.java:56)
	at org.eclipse.jetty.ee10.servlet.ServletContextResponse.write(ServletContextResponse.java:288)
	at org.eclipse.jetty.server.handler.ErrorHandler.generateAcceptableResponse(ErrorHandler.java:254)
	at org.eclipse.jetty.server.handler.ErrorHandler.generateResponse(ErrorHandler.java:160)
	at org.eclipse.jetty.server.handler.ErrorHandler.handle(ErrorHandler.java:115)
	at org.eclipse.jetty.ee10.servlet.ServletChannel.handle(ServletChannel.java:496)
	at org.eclipse.jetty.ee10.servlet.ServletHandler.handle(ServletHandler.java:464)
	at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:597)
	at org.eclipse.jetty.server.handler.ContextHandler.handle(ContextHandler.java:851)
	at org.eclipse.jetty.server.Handler$Wrapper.handle(Handler.java:740)
	at org.eclipse.jetty.server.handler.EventsHandler.handle(EventsHandler.java:81)
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:151)
	at org.eclipse.jetty.server.Handler$Wrapper.handle(Handler.java:740)
	at org.eclipse.jetty.server.handler.EventsHandler.handle(EventsHandler.java:81)
	at org.eclipse.jetty.server.Server.handle(Server.java:179)
	at org.eclipse.jetty.server.internal.HttpChannelState$HandlerInvoker.run(HttpChannelState.java:635)
	at org.eclipse.jetty.util.thread.Invocable$ReadyTask.run(Invocable.java:105)
	at org.eclipse.jetty.http2.server.internal.HttpStreamOverHTTP2$1.run(HttpStreamOverHTTP2.java:133)
	at org.eclipse.jetty.util.thread.MonitoredQueuedThreadPool$1.run(MonitoredQueuedThreadPool.java:73)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:979)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1209)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1164)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

Since that exception is thrown from the lowest layer, it is the `HTTP2Flusher` that catches it and its only option is to close the TCP connection as it cannot differentiate between streams.

**Comment by gregw** [comment]
@sbordet @lorban I think we could use the callback below.   The `onFailure` and `onIdleTimeout` methods would cancel the callback, which if of this type would separate out the failure from the completion.    But the issue may be, what if a CancellableCallback is wrapped by another callback?  Maybe cancellation needs to be a first class method in Callback, so that it can be propagated via wrappers etc.

```java
    /**
     * A Callback that can be cancelled.
     * The callback is cancelled when it is known that it cannot be completed successfully, but a scheduled callback to
     * {@link #succeeded()} or {@link #failed(Throwable)} cannot be aborted.
     * The following methods can be extended:
     * <ul>
     *     <li>{@link #onSuccess()} will be called by a call to {@link #succeeded()} if the Callback has not already been
     *     {@link #cancel(Throwable) cancelled}.</li>
     *     <li>{@link #onFailure(Throwable)} will be called by a call to {@link #failed(Throwable)} if the Callback has not
     *     already been {@link #cancel(Throwable) cancelled}.</li>
     *     <li>{@link #onComplete(Throwable)} will always be called once either {@link #succeeded()} or
     *     {@link #failed(Throwable)} has been called.  If passed {@link Throwable} will be null only if the callback has
     *     neither been {@link #cancel(Throwable) cancelled} nor {@link #failed(Throwable) failed}.</li>
     * </ul>
     */
    class CancelableCallback implements Callback
    {
        private static final Throwable SUCCEEDED = new StaticException("Succeeded");
        AtomicMarkableReference<Throwable> _completion = new AtomicMarkableReference<>(null, false);

        /**
         * Cancel or fail a callback.
         * @param callback The {@link Callback} to {@link #cancel(Throwable)} if it is an instance of {@code CancelableCallback},
         *                 else it is {@link #failed(Throwable) failed}.
         * @param cause The cause of the cancellation
         */
        static void cancel(Callback callback, Throwable cause)
        {
            if (callback instanceof CancelableCallback cancelableCallback)
                cancelableCallback.cancel(cause);
            else 
                callback.failed(cause);
        }

        /**
         * Cancel the callback if it has not already been completed.
         * The {@link #onFailure(Throwable)} method will be called directly by this call, then
         * the {@link #onComplete(Throwable)} method will be called only once the {@link #succeeded()} or
         * {@link #failed(Throwable)} methods are called.
         * @param cause The cause of the cancellation
         */
        public void cancel(Throwable cause)
        {
            if (cause == null)
                cause = new CancellationException();
            if (_completion.compareAndSet(null, cause, false, false))
                onFailure(cause);
        }

        @Override
        public final void failed(Throwable failure)
        {
            if (failure == null)
                failure = new Exception();
            while (true)
            {
                if (_completion.compareAndSet(null, failure, false, true))
                {
                    try
                    {
                        onFailure(failure);
                    }
                    finally
                    {
                        onComplete(null);
                    }
                    return;
                }

                if (_completion.isMarked())
                    return;

                Throwable cause = _completion.getReference();
                ExceptionUtil.addSuppressedIfNotAssociated(cause, failure);
                if (_completion.compareAndSet(cause, cause, false, true))
                {
                    onComplete(cause);
                    return;
                }
            }
        }

        @Override
        public final void succeeded()
        {
            while (true)
            {
                if (_completion.compareAndSet(null, SUCCEEDED, false, true))
                {
                    try
                    {
                        onSuccess();
                    }
                    finally
                    {
                        onComplete(null);
                    }
                    return;
                }

                if (_completion.isMarked())
                    return;

                Throwable cause = _completion.getReference();
                if (_completion.compareAndSet(cause, cause, false, true))
                {
                    onComplete(cause);
                    return;
                }
            }
        }

        /**
         * Called when the callback has been {@link #succeeded() succeeded} but not {@link #cancel(Throwable) cancelled}.
         * The {@link #onComplete(Throwable)} method will be called with a {@code null} argument after this call.
         * Typically, this method is implement to act on the success.  It can release or reuse any resources that may have 
         * been in use by the scheduled operation, but it should defer that release or reuse to the subsequent call to
         * {@link #onComplete(Throwable)} to avoid double releasing.
         */
        protected void onSuccess()
        {

        }

        /**
         * Called when the callback has either been {@link #failed(Throwable) failed} or {@link #cancel(Throwable) cancelled}.
         * The {@link #onComplete(Throwable)} method will ultimately be called, but only once the callback has been
         * {@link #succeeded() succeeded} or {@link #failed(Throwable)}.
         * Typically, this method is implemented to act on the failure, but it cannot release or reuse any resources that may
         * be in use by the schedule operation.
         * @param cause The cause of the failure or cancellation
         */
        protected void onFailure(Throwable cause)
        {

        }

        /**
         * Called once the callback has been either {@link #succeeded() succeeded} or {@link #failed(Throwable)}.
         * Typically, this method is implemented to release resources that may be used by the scheduled operation.
         * @param cause The cause of the failure or cancellation
         */
        protected void onComplete(Throwable cause)
        {

        }
    }
```

**Comment by gregw** [comment]
thought bubble for this is in #11857.  A fair bit more work is needed... and I'm not entirely convinced it is doable yet.

**Comment by gregw** [comment]
@lorban I agree this approach is invasive... and I have yet to work it all the way through to even show that it is feasible.   But I already don't think your proposal in #11860 is feasible (see comments there).

What I'd like to do is for us to collaborate on this PR until it is either working... or has discovered all the issues that we don't know about yet from the 20,000 foot view.

Specifically, could you have a go at updating the HttpOutput callbacks so that they separate out failure from completion?  I'm not entirely sure that can be done with this proposal... or that the contract is exactly correct if it is. 
I will keep looking at all our callback wrappers and see how (if?) cancel semantics can be applied to them.

Once we have succeeded in making this PR work (or found the killer issue that prevents it from working), then we will be in a position to evaluate if it really is the way to go... brain storm alternatives... look for simplifications.. etc.

**Comment by lorban** [comment]
@gregw As a summary, we are in agreement that the problem is that we need a formal way to asynchronously cancel callbacks, and that every place where such async cancelling happens should be modified. What we're unsure about is the way to implement that, as `Callback` is central to the way Jetty works so we need to come with a robust new mechanism while also keep the changes to a minimum and stay backward compatible.

I'm going to scavenge through the code base to try to identify all the places where an async cancellation is needed and modify them to separate failure from completion. `HttpOutput`s, `ErrorHandler` and `IteratingCallback` are impacted but there's probably more.

**Comment by gregw** [comment]
@lorban @sbordet I've closed #11857 as a dead end, as my second attempt also got bogged down.

However, I am going to have another go as I think there is still some hope.     I think `IteratingCallback` is key, as: 
  a) It is frequently used on our `Callback` wrapping chains
  b) It already has `abort` semantic.
  c) I can see issues with it's `abort` implementation, as it doesn't really serialize the abort call.
  d) The ICB `onCompleteSuccess()` and `onCompleteFailure(Throwable)` methods already give us a vocab and concepts we can build on.

Let's hangout tonight (my time) and discuss
